### PR TITLE
Html non secure permissions

### PR DESCRIPTION
**Pull request by ekuleshov**
This should fix https://github.com/Baseflow/flutter-permission-handler/issues/1239

## Pre-launch Checklist

- [x] I made sure the project builds.
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is does not need version changes.
- [x] I updated `CHANGELOG.md` to add a description of the change.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I rebased onto `main`.
- [ ] I added new tests to check the change I am making, or this PR does not need tests.
- [x] I made sure all existing and new tests are passing.
- [x] I ran `dart format .` and committed any changes.
- [x] I ran `flutter analyze` and fixed any errors.

[Contributor Guide]: https://github.com/Baseflow/flutter-permission-handler/blob/master/CONTRIBUTING.md
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

**Comment by TimHoogstrate** [comment]
Dear @ekuleshov,

I saw your issue and the solution seems to be fine. However, I am not able to test the PR at the moment and I lack the required knowledge about the topic. Could you explain why this is happening and why it is now resolved?

Kind regards,

**Review by TimHoogstrate — COMMENTED** [review]
Code LGTM

**Comment by ekuleshov** [comment]
> I saw your issue and the solution seems to be fine. However, I am not able to test the PR at the moment and I lack the required knowledge about the topic. Could you explain why this is happening and why it is now resolved?

This issue is actually pretty bad... As soon as you add the current version of `flutter-permission-handler` to the project yaml or even add some other plugin that has `flutter-permission-handler` as a transitive dependency, the web app fail to start when launched from a non-https location (in our case it was an internal web app and https is hard to setup).

From HTML docs for the [mediaDevices](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/mediaDevices) property:

> Secure context: This feature is available only in [secure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts) (HTTPS), in some or all [supporting browsers](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/mediaDevices#browser_compatibility).

For some reason the `WebPermissionHandler` is accessing the `html.window.navigator.mediaDevices` in the static field initializer and forcing value as non-null. That is where it throws a null pointer exception, which is not easy to see, unless you launch a non obfuscated web app in debugger with `--web-hostname=<host ip address>` option using non-localhost ip address.

**Comment by mvanbeusekom** [comment]
Hi @ekuleshov,

Thank you for this PR, upon request from @TimHoogstrate I did another code review. 

First I think you are correct that the static `_devices` field should be optional as per your explanation and supplied fixes. This however will not completely resolve the issue as the problem will now be moved down the line into the `WebDelegate` class. An `Unexpected null value` exception will now be thrown when someone is trying to request camera or microphone permissions as they rely on the `devices` variable not to be `null` (which is a design issue that should be fixed).

It would be great if you could update this PR and have the `_requestCameraPermission` and `_requestMicrophonePermission` methods accept an nullable instance of `html.MediaDevices` as a parameter and have them `return false;` when the `devices` parameter is `null`. 

If you don't have the time please let us know, in that case we would be happy to take over the PR and add the necessary changes.

File: permission_handler_html/lib/web_delegate.dart [hunk 1]
```diff
@@ -77,8 +74,8 @@ class WebDelegate {
       // The manual stop action is then needed here for to stop the automatic
       // recording.
 
-      if (mediaStream.active!) {
-        final audioTracks = mediaStream.getAudioTracks();
+      if (mediaStream?.active ?? false) {
+        final audioTracks = mediaStream!.getAudioTracks();
         if (audioTracks.isNotEmpty) {
           audioTracks[0].stop();
         }
```
Comment (mvanbeusekom) [inline permission_handler_html/lib/web_delegate.dart:81]: Combining the condition here mixes two different situations and treating them the same, which leads to unexpected behaviour.

1. The original test is just to make sure the `mediaStream` is stopped if it is started (which is caused by calling the `getUserMedia()` method on line 68).
2. The added tests checks if `mediaStream` is `null`, which is most likely the result of `_devices` being `null`. 

I feel we should `return false;` at the start of the method if `_devices` is `null` and prevent this mixed behaviour. Added benefit will be that returning `false` is translated to a `PermissionStatus.denied` which is was it comes down to in a HTTP (non-secure) context.

File: permission_handler_html/lib/web_delegate.dart [hunk 2]
```diff
@@ -103,8 +98,8 @@ class WebDelegate {
       // The manual stop action is then needed here for to stop the automatic
       // recording.
 
-      if (mediaStream.active!) {
-        final videoTracks = mediaStream.getVideoTracks();
+      if (mediaStream?.active ?? false) {
```
Comment (mvanbeusekom) [inline permission_handler_html/lib/web_delegate.dart:101]: As with the microphone permissions, we should `return false;` earlier in my opinion

**Review by mvanbeusekom — CHANGES_REQUESTED** [review]

File: permission_handler_html/lib/web_delegate.dart (hunk 1 above)
Comment (ekuleshov) [reply permission_handler_html/lib/web_delegate.dart]: > 2\. The added tests checks if `mediaStream` is `null`, which is most likely the result of `_devices` being `null`

Not only. The `mediaStream.active` value is also nullable. Though a good catch on `PermissionStatus.denied`. 

The original behaviour to return `true` by default is now how I'd do it, but I didn't want to completely rewrite this. Perhaps someone could improve it when doing refactoring to remove dependency on `html` packages in order to support wasm...

**Review by mvanbeusekom — APPROVED** [review]
LGTM!

Thank you for the additional testing and of course contributing with these fixes. It is really appreciated.

**Pull request merged**